admin/category_product_listing.php

```diff
@@ -25,1230 +25,1656 @@
-      zen_update_attributes_products_option_values_sort_order($_GET['products_id']);
-      $messageStack->add_session(SUCCESS_ATTRIBUTES_UPDATE . ' ID#' . $_GET['products_id'], 'success');
-      $action = '';
-      zen_redirect(zen_href_link(FILENAME_CATEGORY_PRODUCT_LISTING, 'cPath=' . $cPath . '&pID=' . $_GET['products_id'] . (isset($_GET['page']) ? '&page=' . $_GET['page'] : '')));
-      break;
-
-    // attributes copy to product
-    case 'update_attributes_copy_to_product':
-      $copy_attributes_delete_first = ($_POST['copy_attributes'] === 'copy_attributes_delete' ? '1' : '0');
-      $copy_attributes_duplicates_skipped = ($_POST['copy_attributes'] === 'copy_attributes_ignore' ? '1' : '0');
-      $copy_attributes_duplicates_overwrite = ($_POST['copy_attributes'] === 'copy_attributes_update' ? '1' : '0');
-      zen_copy_products_attributes($_POST['products_id'], $_POST['products_update_id']);
-      //      die('I would copy Product ID#' . $_POST['products_id'] . ' to a Product ID#' . $_POST['products_update_id'] . ' - Existing attributes ' . $_POST['copy_attributes']);
-      $_GET['action'] = '';
-      zen_redirect(zen_href_link(FILENAME_CATEGORY_PRODUCT_LISTING, 'cPath=' . $cPath . '&pID=' . $_GET['products_id'] . (isset($_GET['page']) ? '&page=' . $_GET['page'] : '')));
-      break;
-
-    // attributes copy to category
-    case 'update_attributes_copy_to_category':
-      $copy_attributes_delete_first = ($_POST['copy_attributes'] === 'copy_attributes_delete' ? '1' : '0');
-      $copy_attributes_duplicates_skipped = ($_POST['copy_attributes'] === 'copy_attributes_ignore' ? '1' : '0');
-      $copy_attributes_duplicates_overwrite = ($_POST['copy_attributes'] === 'copy_attributes_update' ? '1' : '0');
-      $copy_to_category = zen_get_linked_products_for_category((int)$_POST['categories_update_id']);
-      foreach ($copy_to_category as $item) {
-        zen_copy_products_attributes($_POST['products_id'], $item);
-      }
-      //      die('CATEGORIES - I would copy Product ID#' . $_POST['products_id'] . ' to a Category ID#' . $_POST['categories_update_id']  . ' - Existing attributes ' . $_POST['copy_attributes'] . ' Total Products ' . $copy_to_category->RecordCount());
-
-      $_GET['action'] = '';
-      zen_redirect(zen_href_link(FILENAME_CATEGORY_PRODUCT_LISTING, 'cPath=' . $cPath . '&pID=' . $_GET['products_id'] . (isset($_GET['page']) ? '&page=' . $_GET['page'] : '')));
-      break;
-    case 'setflag_categories':
-    case 'delete_category':
-    case 'move_category':
-    case 'delete_product':
-    case 'move_product':
-    case 'copy_product':
-    case 'attribute_features':
-    case 'attribute_features_copy_to_product':
-    case 'attribute_features_copy_to_category':
-      break;
-    default:
-      $zco_notifier->notify('NOTIFY_ADMIN_PROD_LISTING_DEFAULT_ACTION');
-      $action = $_GET['action'] = '';
-      break;
-  }
+                        WHERE categories_id = " . (int)$categories_id
```

Comment:
$categories_id is already 'known' to be an int.

Reply:
I seem to recall a discussion about these unnecessary castings in sql being retained as a bit of a safety belt-and-braces approach. Usually the IDE points out these surplus castings: comes under Remove surplus sanitation: https://github.com/zencart/documentation/issues/879

For next PR.